### PR TITLE
Support a new translator API: DeepL

### PR DESCRIPTION
**Pull request by naoyukik**
Implemented DeepL

Notes
- I do not have an account with DeepL Pro. Therefore, I do not know if it is working.
  - I would expect the API to work since it's just a different URL.
- Untranslated messages in Chinese and Korean.
- Glossary API is not implemented.

refs: #472

### CODE REVIEW

**Comment by YiiGuxing** [comment]
Thank you very much for your PR!
I'll find time to review it, it may not be merged until some later version.

**Comment by YiiGuxing** [comment]
@naoyukik Hi,
There is currently a conflict between the branches, can you help to resolve it? You just need to commit normally after resolving it in your branch.

Thanks.

**Comment by naoyukik** [comment]
@YiiGuxing 
Thanks for the reply.
Conflict resolved！

**Comment by YiiGuxing** [comment]
Thanks.

**Comment by naoyukik** [comment]
@YiiGuxing 
Document translation implementation was omitted and has been added.
I will hold off on the implementation regarding DeepL until this is merged.

Thank you.

**Comment by YiiGuxing** [comment]
@naoyukik 
Thanks, 
In fact you don't need to implement the `DocumentationTranslator` interface unless it has a dedicated documentation translation API.

I'm currently working on the release of version 3.3, until then, I don't have time for this PR.

**Comment by naoyukik** [comment]
@YiiGuxing 

> I'm currently working on the release of version 3.3, until then, I don't have time for this PR.

Understood. No problem.


> In fact you don't need to implement the DocumentationTranslator interface unless it has a dedicated documentation translation API.

DeepL has the ability to read xml and html tags and translate the non-tag portions depending on the configuration.

The following commit implements it.
[feat: Implement process for document translation](https://github.com/YiiGuxing/TranslationPlugin/pull/986/commits/171cd369c2dbf05624c90cc9e6bf952b332324fa)

Currently, I believe this is sufficient to handle document translation.
I have checked that it works, and the documents are being translated while maintaining their formatting.

Or are there any implementation requirements for the "document translation" function itself?

Please reply when you have time available.

Thank you.

**Comment by YiiGuxing** [comment]
> > In fact you don't need to implement the DocumentationTranslator interface unless it has a dedicated documentation translation API.
> 
> DeepL has the ability to read xml and html tags and translate the non-tag portions depending on the configuration.
> 
> The following commit implements it.
> [feat: Implement process for document translation](https://github.com/YiiGuxing/TranslationPlugin/pull/986/commits/171cd369c2dbf05624c90cc9e6bf952b332324fa)
> 
> Currently, I believe this is sufficient to handle document translation.
> I have checked that it works, and the documents are being translated while maintaining their formatting.
> 
> Or are there any implementation requirements for the "document translation" function itself?

@naoyukik hi,
I saw the relevant documentation. But it should be only document translation handling tags, normal translation does not need to be handled.

**Comment by naoyukik** [comment]
@YiiGuxing 
Hmmm...I will check the source code again.

**Comment by naoyukik** [comment]
I checked the source code.

The normal translation process would be [documentation/HtmlTranslator.kt:128](https://github.com/YiiGuxing/TranslationPlugin/blob/master/src/main/kotlin/cn/yiiguxing/plugin/translate/documentation/HtmlTranslator.kt#L128) The html tag is trimmed in

For document translation, we want to leave the html tags as they are, so we implement document translation (`DocumentationTranslator`).

In the meantime, please let me know if there are any mistakes when you check the source code here.

**Comment by YiiGuxing** [comment]
I understand what you mean, 
but what I mean is that the `tag_handling` parameter is only passed in when translating documents, and we should not pass this parameter in normal translation.

Also, I noticed that you did not use the `srcLang` parameter when requesting the translation API:
https://github.com/YiiGuxing/TranslationPlugin/blob/171cd369c2dbf05624c90cc9e6bf952b332324fa/src/main/kotlin/cn/yiiguxing/plugin/translate/trans/deepl/DeeplTranslator.kt#L119-L132

This makes the program behave against the user's expectations, so this parameter cannot be ignored, we can ignore it only if the value of `srcLang` is `Lang.AUTO`, `Lang.AUTO` means auto detect language.

**Comment by naoyukik** [comment]
@YiiGuxing 
Thank you for your support.

- Changed "tag_handling" so that it is only output when the document is translated.
- "srcLang" was changed so that "source_lang" is output when it is not AUTO.

File: src/main/kotlin/cn/yiiguxing/plugin/translate/trans/deepl/DeeplTranslator.kt
Comment (YiiGuxing) [inline src/main/kotlin/cn/yiiguxing/plugin/translate/trans/deepl/DeeplTranslator.kt:123]: Using `val isFree = privateKey.endsWith(":fx")` here will be more concise and efficient.

File: src/main/kotlin/cn/yiiguxing/plugin/translate/trans/deepl/DeeplTranslator.kt
Comment (YiiGuxing) [inline src/main/kotlin/cn/yiiguxing/plugin/translate/trans/deepl/DeeplTranslator.kt:136]: Should be `postData["source_lang"] = srcLang.deeplLanguageCode` here.

File: src/main/kotlin/cn/yiiguxing/plugin/translate/trans/deepl/DeeplTranslator.kt
Comment (YiiGuxing) [inline src/main/kotlin/cn/yiiguxing/plugin/translate/trans/deepl/DeeplTranslator.kt:188]: The cache key should be updated before calling the `execute` method to avoid cache key conflicts:
```
SimpleTranslateClient.updateCacheKey { it.update("DOCUMENTATION".toByteArray()) }
```

File: src/main/kotlin/cn/yiiguxing/plugin/translate/trans/AbstractTranslator.kt
Comment (YiiGuxing) [inline src/main/kotlin/cn/yiiguxing/plugin/translate/trans/AbstractTranslator.kt:48]: The code related to the specific translator should not be placed here, it should be implemented in the specific translator.

**Review by YiiGuxing — CHANGES_REQUESTED** [review]

File: src/main/kotlin/cn/yiiguxing/plugin/translate/trans/deepl/DeeplTranslator.kt
Comment (naoyukik) [reply src/main/kotlin/cn/yiiguxing/plugin/translate/trans/deepl/DeeplTranslator.kt]: Thanks. Corrected.

File: src/main/kotlin/cn/yiiguxing/plugin/translate/trans/deepl/DeeplTranslator.kt
Comment (naoyukik) [reply src/main/kotlin/cn/yiiguxing/plugin/translate/trans/deepl/DeeplTranslator.kt]: Thanks. Corrected.

File: src/main/kotlin/cn/yiiguxing/plugin/translate/trans/deepl/DeeplTranslator.kt
Comment (naoyukik) [reply src/main/kotlin/cn/yiiguxing/plugin/translate/trans/deepl/DeeplTranslator.kt]: Thanks. Corrected.

File: src/main/kotlin/cn/yiiguxing/plugin/translate/trans/AbstractTranslator.kt
Comment (naoyukik) [reply src/main/kotlin/cn/yiiguxing/plugin/translate/trans/AbstractTranslator.kt]: Thanks.
This section has been reverted to its previous state and implemented for DeepL.

**Review by YiiGuxing — APPROVED** [review]

**Comment by YiiGuxing** [comment]
merged, thanks!

**Comment by YiiGuxing** [comment]
Your changes are merged into [feature/depp-l](https://github.com/YiiGuxing/TranslationPlugin/tree/feature/depp-l) branch, It will be further refined and merged into the master branch in due course. Thanks again.

**Comment by naoyukik** [comment]
@YiiGuxing 
Thank you for teaching me so patiently. I really appreciate it.

**Comment by YiiGuxing** [comment]
@naoyukik hi,

Thanks again for your implementation of the DeeL translator, I'm now getting ready to merge into the next version: [v3.4](https://github.com/YiiGuxing/TranslationPlugin/milestone/31). However, due to regional reasons, the access token of my DeepL account does not have permission to access the DeepL REST API, so I can't do some optimizations and tests. Could you please provide me with a temporary access token so I can do some testing? I assure you that I will only use it for testing purposes. If you find anything unusual, you can revoke the token at any time:





If you would like, please send the token to my email: yii.guxing@gmail.com.
If you have any concerns, please let me know.

Thank you so much!

**Comment by naoyukik** [comment]
@YiiGuxing 
Sorry, I am not able to give you the API Key.

It will take some time, but if you let us know your policy for the fixes, we will do our best to accommodate you.
Also, if you put the source code into our branch, we will check the code to make sure it works.

I will help you as much as I can until release, so please let me know if you need any help.

**Comment by YiiGuxing** [comment]
@naoyukik hi,
Thank you for your note, it's okay, I understand your concerns. After I finish developing it, I may need your help to test.

**Pull request merged**
